File: code/lua/exercises.lua
first_then_apply = function(seq, pred, func)
  for _, x in ipairs(seq) do
    if pred(x) then
      return func(x)
    end
  end
  return nil
end

powers_generator = function(base, limit)
  return coroutine.create(function()
    local value = 1
    while value <= limit do
      coroutine.yield(value)
      value = value * base
    end
  end)
end

say = function(word)
  if word == nil then
    return ""
  end
  local words = { word }
  local function inner(nextWord)
    if nextWord == nil then
      return table.concat(words, " ")
    end
    table.insert(words, nextWord)
    return inner
  end
  return inner
end

meaningful_line_count = function(filename)
  local file, err = io.open(filename, "r")
  if not file then
    error(err, 2)
  end

  local lines = {}
  for line in file:lines() do
    table.insert(lines, line)
  end
  file:close()

  local count = 0
  for _, line in ipairs(lines) do
    local stripped = line:gsub("^%s*", "")
    if not stripped:match("^#") then
      count = count + 1
    end
  end

  local n = #lines
  if n > 0 then
    local last = lines[n]
    local trimmed_last = last:match("^%s*(.-)%s*$")
    local stripped_last = last:gsub("^%s*", "")
    if trimmed_last == "" and not stripped_last:match("^#") then
      count = count - 1
    end
  end

  return count
end

Quaternion = {}
Quaternion.__index = Quaternion

function Quaternion.new(a, b, c, d)
  local self = setmetatable({}, Quaternion)
  self.a = a
  self.b = b
  self.c = c
  self.d = d
  return self
end

function Quaternion:coefficients()
  return { self.a, self.b, self.c, self.d }
end

function Quaternion:conjugate()
  return Quaternion.new(self.a, -self.b, -self.c, -self.d)
end

function Quaternion.__add(q1, q2)
  return Quaternion.new(
    q1.a + q2.a,
    q1.b + q2.b,
    q1.c + q2.c,
    q1.d + q2.d
  )
end

function Quaternion.__mul(q1, q2)
  local a1, b1, c1, d1 = q1.a, q1.b, q1.c, q1.d
  local a2, b2, c2, d2 = q2.a, q2.b, q2.c, q2.d
  return Quaternion.new(
    a1 * a2 - b1 * b2 - c1 * c2 - d1 * d2,
    a1 * b2 + b1 * a2 + c1 * d2 - d1 * c2,
    a1 * c2 - b1 * d2 + c1 * a2 + d1 * b2,
    a1 * d2 + b1 * c2 - c1 * b2 + d1 * a2
  )
end

function Quaternion.__eq(q1, q2)
  return q1.a == q2.a and q1.b == q2.b and q1.c == q2.c and q1.d == q2.d
end

local function is_nonzero(x)
  return math.abs(x) > 0
end

local function num_to_str(x)
  if x == math.floor(x) then
    return string.format("%.1f", x)
  else
    return tostring(x)
  end
end

function Quaternion:__tostring()
  local a, b, c, d = self.a, self.b, self.c, self.d

  if not (is_nonzero(a) or is_nonzero(b) or is_nonzero(c) or is_nonzero(d)) then
    return "0"
  end

  local parts = {}

  if is_nonzero(a) then
    table.insert(parts, num_to_str(a))
  end

  local function add(coeff, sym)
    if not is_nonzero(coeff) then
      return
    end
    local first = (#parts == 0)
    local sign = coeff < 0 and "-" or "+"
    local mag = math.abs(coeff)
    local mag_str = mag == 1 and "" or num_to_str(mag)

    if first then
      if coeff < 0 then
        table.insert(parts, "-" .. mag_str .. sym)
      else
        table.insert(parts, mag_str .. sym)
      end
    else
      table.insert(parts, sign .. mag_str .. sym)
    end
  end

  add(b, "i")
  add(c, "j")
  add(d, "k")

  return table.concat(parts)
end

File: code/swift/exercises.swift
import Foundation

func firstThenLowerCase(of strings: [String], satisfying predicate: (String) -> Bool) -> String? {
    strings.first(where: predicate)?.lowercased()
}

struct Say {
    private let words: [String]

    func and(_ word: String) -> Say {
        Say(words: words + [word])
    }

    var phrase: String {
        words.joined(separator: " ")
    }

    init(words: [String]) {
        self.words = words
    }
}

func say() -> Say {
    Say(words: [])
}

func say(_ word: String) -> Say {
    Say(words: [word])
}

func meaningfulLineCount(_ path: String) async -> Result<Int, Error> {
    let url = URL(fileURLWithPath: path)
    do {
        let handle = try FileHandle(forReadingFrom: url)
        var count = 0
        defer { try? handle.close() }
        for try await line in handle.bytes.lines {
            if !line.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
                count += 1
            }
        }
        return .success(count)
    } catch {
        return .failure(error)
    }
}

struct Quaternion: Equatable, CustomStringConvertible {
    let a: Double
    let b: Double
    let c: Double
    let d: Double

    init(a: Double = 0, b: Double = 0, c: Double = 0, d: Double = 0) {
        self.a = a
        self.b = b
        self.c = c
        self.d = d
    }

    var coefficients: [Double] {
        [a, b, c, d]
    }

    var conjugate: Quaternion {
        Quaternion(a: a, b: -b, c: -c, d: -d)
    }

    static let ZERO = Quaternion(a: 0, b: 0, c: 0, d: 0)
    static let I = Quaternion(a: 0, b: 1, c: 0, d: 0)
    static let J = Quaternion(a: 0, b: 0, c: 1, d: 0)
    static let K = Quaternion(a: 0, b: 0, c: 0, d: 1)

    static func + (lhs: Quaternion, rhs: Quaternion) -> Quaternion {
        Quaternion(
            a: lhs.a + rhs.a,
            b: lhs.b + rhs.b,
            c: lhs.c + rhs.c,
            d: lhs.d + rhs.d
        )
    }

    static func * (lhs: Quaternion, rhs: Quaternion) -> Quaternion {
        let a1 = lhs.a, b1 = lhs.b, c1 = lhs.c, d1 = lhs.d
        let a2 = rhs.a, b2 = rhs.b, c2 = rhs.c, d2 = rhs.d
        return Quaternion(
            a: a1 * a2 - b1 * b2 - c1 * c2 - d1 * d2,
            b: a1 * b2 + b1 * a2 + c1 * d2 - d1 * c2,
            c: a1 * c2 - b1 * d2 + c1 * a2 + d1 * b2,
            d: a1 * d2 + b1 * c2 - c1 * b2 + d1 * a2
        )
    }

    var description: String {
        func nonZero(_ x: Double) -> Bool { abs(x) > 0.0 }

        if !nonZero(a) && !nonZero(b) && !nonZero(c) && !nonZero(d) {
            return "0"
        }

        var s = ""

        if nonZero(a) {
            s.append(String(a))
        }

        func appendTerm(_ coeff: Double, _ symbol: String) {
            if !nonZero(coeff) { return }
            let first = s.isEmpty
            let mag = abs(coeff)
            let magStr = mag == 1.0 ? "" : String(mag)
            if first {
                if coeff < 0 {
                    s.append("-" + magStr + symbol)
                } else {
                    s.append(magStr + symbol)
                }
            } else {
                let sign = coeff < 0 ? "-" : "+"
                s.append(sign + magStr + symbol)
            }
        }

        appendTerm(b, "i")
        appendTerm(c, "j")
        appendTerm(d, "k")

        return s
    }
}

indirect enum BinarySearchTree {
    case empty
    case node(left: BinarySearchTree, value: String, right: BinarySearchTree)

    var size: Int {
        switch self {
        case .empty:
            return 0
        case let .node(left, _, right):
            return left.size + 1 + right.size
        }
    }

    func contains(_ v: String) -> Bool {
        switch self {
        case .empty:
            return false
        case let .node(left, value, right):
            if v == value { return true }
            if v < value { return left.contains(v) }
            return right.contains(v)
        }
    }

    func insert(_ v: String) -> BinarySearchTree {
        switch self {
        case .empty:
            return .node(left: .empty, value: v, right: .empty)
        case let .node(left, value, right):
            if v == value { return self }
            if v < value {
                return .node(left: left.insert(v), value: value, right: right)
            } else {
                return .node(left: left, value: value, right: right.insert(v))
            }
        }
    }
}

extension BinarySearchTree: CustomStringConvertible {
    var description: String {
        switch self {
        case .empty:
            return "()"
        case let .node(left, value, right):
            let leftStr: String = {
                if case .empty = left { return "" }
                return left.description
            }()
            let rightStr: String = {
                if case .empty = right { return "" }
                return right.description
            }()
            return "(\(leftStr)\(value)\(rightStr))"
        }
    }
}
